sui_programmability/examples/defi/sources/pool.move

```diff
@@ -0,0 +1,530 @@
+        _: P,
+        token: Coin<T>,
+        sui: Coin<SUI>,
+        fee_percent: u64,
+        ctx: &mut TxContext
+    ): Coin<LSP<P, T>> {
+        let sui_amt = coin::value(&sui);
+        let tok_amt = coin::value(&token);
+
+        assert!(sui_amt > 0 && tok_amt > 0, EZeroAmount);
+        assert!(sui_amt < MAX_POOL_VALUE && tok_amt < MAX_POOL_VALUE, EPoolFull);
+        assert!(fee_percent >= 0 && fee_percent < 10000, EWrongFee);
+
+        // Initial share of LSP is the sqrt(a) * sqrt(b)
+        let share = math::sqrt(sui_amt) * math::sqrt(tok_amt);
+        let lsp_treasury = coin::create_currency(LSP<P, T> {}, ctx);
+        let lsp = coin::mint(&mut lsp_treasury, share, ctx);
+
+        transfer::share_object(Pool {
+            info: object::new(ctx),
+            token: coin::into_balance(token),
+            sui: coin::into_balance(sui),
+            lsp_treasury,
+            fee_percent
+        });
+
+        lsp
+    }
+
+
+    /// Entrypoint for the `swap_sui` method. Sends swapped token
+    /// to sender.
+    entry fun swap_sui_<P, T>(
+        pool: &mut Pool<P, T>, sui: Coin<SUI>, ctx: &mut TxContext
+    ) {
+        transfer::transfer(
+            swap_sui(pool, sui, ctx),
+            tx_context::sender(ctx)
+        )
+    }
+
+    /// Swap `Coin<SUI>` for the `Coin<T>`.
+    /// Returns Coin<T>.
+    public fun swap_sui<P, T>(
+        pool: &mut Pool<P, T>, sui: Coin<SUI>, ctx: &mut TxContext
+    ): Coin<T> {
+        assert!(coin::value(&sui) > 0, EZeroAmount);
+
+        let sui_balance = coin::into_balance(sui);
+
+        // Calculate the output amount - fee
+        let (sui_reserve, token_reserve, _) = get_amounts(pool);
+        let output_amount = get_input_price(
+            balance::value(&sui_balance),
+            sui_reserve,
+            token_reserve,
+            pool.fee_percent
+        );
+
+        balance::join(&mut pool.sui, sui_balance);
+        coin::take(&mut pool.token, output_amount, ctx)
+    }
+
+    /// Entry point for the `swap_token` method. Sends swapped SUI
+    /// to the sender.
+    entry fun swap_token_<P, T>(
+        pool: &mut Pool<P, T>, token: Coin<T>, ctx: &mut TxContext
+    ) {
+        transfer::transfer(
+            swap_token(pool, token, ctx),
+            tx_context::sender(ctx)
+        )
+    }
+
+    /// Swap `Coin<T>` for the `Coin<SUI>`.
+    /// Returns the swapped `Coin<SUI>`.
+    public fun swap_token<P, T>(
+        pool: &mut Pool<P, T>, token: Coin<T>, ctx: &mut TxContext
+    ): Coin<SUI> {
+        assert!(coin::value(&token) > 0, EZeroAmount);
+
+        let tok_balance = coin::into_balance(token);
+        let (sui_reserve, token_reserve, _) = get_amounts(pool);
+        let output_amount = get_input_price(
+            balance::value(&tok_balance),
+            token_reserve,
+            sui_reserve,
+            pool.fee_percent
+        );
+
+        assert!(sui_reserve > 0 && token_reserve > 0, EReservesEmpty);
```

Comment:
Should this be`>=`? Worried that the last token in the pool can't be drained otherwise.